From d7240323b2dcb20a4502f4d2f63b54a06cd98c92 Mon Sep 17 00:00:00 2001
From: Alona Solntseva <alona.solntseva@tandemg.com>
Date: Wed, 16 Feb 2022 11:39:48 +0200
Subject: [PATCH] [SW-67906] gaudi2: erom pcie: separate ar0 code

Create separate secondary_main code and arc0 code for erom
Return ppboot secondary_main code to use preboot_pcie_monitoring instead
of pcie_config
erom secondary_main initialize the pci to gen1
ppboot secondary_main initialize the pci to gen3

Signed-off-by: Alona Solntseva <alona.solntseva@tandemg.com>
Change-Id: I27b89a5f0cdfaa12ce091713092b3b1de0d8fe2a
---

diff --git a/app/ext_rom/Makefile b/app/ext_rom/Makefile
index 9252afe..fd1fa95 100644
--- a/app/ext_rom/Makefile
+++ b/app/ext_rom/Makefile
@@ -41,7 +41,7 @@
 	$(HL_DEV_ARCH)/../common/$(HL_TARGET).o \
 	$(HL_DEV_ARCH)/../common/common.o \
 	$(HL_DEV_ARCH)/common.o \
-	$(HL_DEV_ARCH)/secondary.o
+	erom_secondary.o
 
 EROM_RECOVERY_DEPS := \
 	recovery_stubs.o \
diff --git a/app/ext_rom/erom_secondary.S b/app/ext_rom/erom_secondary.S
new file mode 100644
index 0000000..bfc9c6d
--- /dev/null
+++ b/app/ext_rom/erom_secondary.S
@@ -0,0 +1,18 @@
+/* SPDX-License-Identifier:     GPL-2.0+ */
+/*
+ * Copyright (C) 2021 HabanaLabs Ltd.
+ * All Rights Reserved.
+ *
+ */
+
+.global secondary_entry
+secondary_entry:
+	/* stack top set to secondary CPU DCCM 1k offset (internal address) */
+	mov	sp, 0x80000400
+	mov	fp, sp
+
+	/* Disable un-aligned checks on data memory */
+	lr r0, [STATUS32]
+	bset r0, r0, 0x13
+	kflag r0
+	bl	erom_secondary_main
diff --git a/arch/gaudi2/arc/common.c b/arch/gaudi2/arc/common.c
index cff35e4..350c4d8 100644
--- a/arch/gaudi2/arc/common.c
+++ b/arch/gaudi2/arc/common.c
@@ -630,7 +630,7 @@
 	}
 }
 
-void secondary_main(void)
+void erom_secondary_main(void)
 {
 	pcie_config();
 
@@ -638,6 +638,14 @@
 	sleep();
 }
 
+void secondary_main(void)
+{
+	preboot_pcie_monitoring();
+
+	/* Put secondary CPU in sleep state until further notice */
+	sleep();
+}
+
 void preboot_pcie_wait_for_link(void)
 {
 	u32 *secondary_iccm_addr = (u32 *) ARC0_ICCM_BASE;
